### PR TITLE
Add Optimizer Design Feature Document

### PR DESCRIPTION
**Pull request by a7ehuo**
Authored-By: Vijay Sundaresan <vijaysun@ca.ibm.com>

Signed-off-by: Annabelle Huo <Annabelle.Huo@ibm.com>

### CODE REVIEW

**Comment by a7ehuo** [comment]
@vijaysun-omr @hzongaro May I ask you to review this PR? Thank you!
@0xdaryl fyi

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (vijaysun-omr) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:130]: I believe this refers to the generic apis to get value/block profiling information rather than JProfiling specifically since IProfiling can be a source of profiling information too. This can be seen in consumers such as https://github.com/eclipse/omr/blob/cf8ddbd1adcd87e388cad6be1fd0e0c085285a29/compiler/optimizer/LoopVersioner.cpp#L2726 where the profiling information is obtained without awareness of where it comes from.

**Review by vijaysun-omr — APPROVED** [review]
I don't have major edits to suggest (for obvious reasons) apart from the link correction I mentioned in my prior comment on this PR>

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (dsouzai) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:304]:
```suggestion
The threading and GC components also assume a cooperative model which makes it easier to implement this solution
```

**Review by dsouzai — COMMENTED** [review]
Most of the suggested changes revolve around the fact that this doc will sit in the OpenJ9 repo, and so should remain consistent wrt to keeping the context as OpenJ9.

**Comment by a7ehuo** [comment]
@dsouzai @vijaysun-omr Addressed all the above comments. Ready for another review. Thanks!

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:61]: Some of the information in this section looks like it overlaps with information in the [Compilation Control](https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/control/CompilationControl.md) documentation.  I wonder if it should be consolidated with that section to avoid redundancy, and to make it less likely that one will get out-of-date if the other changes.

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:78]: I wonder whether it would be better to use the term "mode" rather than "model" here.

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:80]: I would suggest changing "since it is better to allow work to be done" to "as that allows work to continue to be done".

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:89]: I'm having trouble parsing this sentence.  Maybe it should read "Performance is not only about peak throughput; other metrics such as. . . ."

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:99]: Should this include a reference to https://github.com/eclipse-openj9/openj9/blob/24b14b4743f82a648712bdca71a5b870828e65f9/doc/compiler/hcr/NextGenHCR.md?

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:105]: "compile time"->"compile-time"

**Review by hzongaro — COMMENTED** [review]
Thanks for pulling this together, Annabelle!  I'm still reviewing, but I'm at a point where it's natural to pause, so I thought I'd post my comments thus far.

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [reply doc/compiler/optimizer/OptimizerDesignFeatures.md:61]: Just to be clear, by "this section", I meant the Heuristics section.

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:215]: "compile time savings" -> "compile-time savings"

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:212]: I'd suggest changing "maintain/recompute" to simply "recompute"

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:249]: "compile-time constraints"

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:247]: "(Directed Acyclic Graph) DAG" -> "Directed Acyclic Graph (DAG)"

Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:247]: A link to [Testarossa's Intermediate Language: An Intro to Trees](https://github.com/eclipse/omr/blob/master/doc/compiler/il/IntroToTrees.md) might be appropriate here.

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:271]: Would a link to the OMR [Data-Flow Engine](https://github.com/eclipse/omr/blob/master/doc/compiler/optimizer/DataFlowEngine.md) be appropriate here?

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:278]: "Meta data"->"Metadata" or maybe "Meta-data"

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:280]: I'd suggest changing "often have other requirements such as support" to something like "often have runtime environments that require support" or maybe simply "often require support"

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:293]: "locate the meta data" -> "locating the metadata"

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:294]: Maybe this would be easier to read as a bulleted list rather than using "a), b), c)"

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:283]: I suspect that it might not be obvious to a reader who has little familiarity with our implementation why exception stack traces and debug code would require state to be updated.  Maybe it would be better to say, 'a need to cope with a change of "state" or capture information about the state of the program'

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:297]: This section uses the word state in scare quotes repeatedly, which suggests that it's not right term.  I wonder whether there's a better term that could be used.  I can't think of one.  If there is no better term, maybe the quotes can be removed.

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (hzongaro) [inline doc/compiler/optimizer/OptimizerDesignFeatures.md:300]: "meta data" -> "metadata" or "meta-data"

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (a7ehuo) [reply doc/compiler/optimizer/OptimizerDesignFeatures.md:30]: I'll update according to the IBM's terminology guidelines:
If they are used as nouns => "compile time" and "run time
If they are used as adjectives => "compile-time" and "runtime"

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (a7ehuo) [reply doc/compiler/optimizer/OptimizerDesignFeatures.md:78]: I think it should be "mode": asynchronous mode, synchronous mode etc.

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (a7ehuo) [reply doc/compiler/optimizer/OptimizerDesignFeatures.md:99]: Good point! I'll include a link pointing to the OSR section: https://github.com/eclipse-openj9/openj9/blob/24b14b4743f82a648712bdca71a5b870828e65f9/doc/compiler/hcr/NextGenHCR.md#osr-implementation-details

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (a7ehuo) [reply doc/compiler/optimizer/OptimizerDesignFeatures.md:247]: I'll add the hyperlink down to where "IL representation" is mentioned again. There is a bookmark in the beginning of the article to jump to this section using `#il-representation`. Not sure what format to use if the section title has a a hyperlink

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (a7ehuo) [reply doc/compiler/optimizer/OptimizerDesignFeatures.md:61]: I think I should link the key concepts mentioned in this Heuristics section to [CompilationControl.md](https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/control/CompilationControl.md) such as invocation counts, compilation threads, Sampler Thread, etc.

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (a7ehuo) [reply doc/compiler/optimizer/OptimizerDesignFeatures.md:176]: My understanding is that these three infrastructures are essential for optimizations based on class type and the fallback during the run time is crucial to make it work

@vijaysun-omr Could you clarify regarding to the above question?

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (a7ehuo) [reply doc/compiler/optimizer/OptimizerDesignFeatures.md:297]: My understanding is that it refers to the program state. I can't think of other words to replace it. 

@vijaysun-omr Is there a better word for it? Or "state" is the accurate word for it?

File: doc/compiler/optimizer/OptimizerDesignFeatures.md
Comment (vijaysun-omr) [reply doc/compiler/optimizer/OptimizerDesignFeatures.md:176]: @a7ehuo your explanation is in line with what I would have said.

**Comment by a7ehuo** [comment]
@hzongaro @vijaysun-omr Addressed the above comments in 9f3cd84. Ready for another review! Thanks!

**Comment by a7ehuo** [comment]
@hzongaro @vijaysun-omr The typos are are fixed in 561c7dc. Ready for another review. Thanks!

**Comment by vijaysun-omr** [comment]
I approve this PR (already approved formally a few days back) after your last edits. Maybe the title as it appears in the table of contents should be something more meaningful, e.g. "Dynamic optimization design philosophy" or "Dynamic optimization design" but I'll leave it up to you.

**Comment by a7ehuo** [comment]
@vijaysun-omr I updated the title in TOC to "Dynamic Optimization Design" 67b73f1

**Comment by hzongaro** [comment]
> I'll merge it once @hzongaro comments on your typo fixes etc.

Thanks, Annabelle @a7ehuo.  Looks good.

Vijay @vijaysun-omr, I think it looks good to merge.  Thanks!

**Pull request merged**
